### ISSUE TITLE
Cuda graph support for FSDP2 is lacking.

### ISSUE THREAD

**Issue by galv** [pytorch/pytorch#164264]
### 🚀 The feature, motivation and pitch

I was asked recently to investigate allowing use of FSDP2 with cuda graph in pytorch.

I discovered several issues:

1. torch.chunk_cat, which is used in FSDP2, allocates pinned host memory. That is currently unsafe in cuda graph capture, but I have a fix to this that I will upload shortly.
2. make_graphed_callables does not support running with models with any hooks. It says you can add hooks to a torch.nn.Module *after* having called make_graphed_callables on it, but this is clumsy because you need the hooks to all-gather all of your weights in order to warmup. Warmup is the only safe way to guarantee a  workload can be captured to a cuda graph today. If you could somehow add the hooks for the sake of warmup, remove the hooks for the sake of make_graphed_callables, and then re-add the hooks, I suppose you could warmup successfully. However, make_graphed_callables is an old API that came out when cuda graphs were first added to pytorch. I personally do not think that [the requirement that a module not have hooks](https://github.com/pytorch/pytorch/blob/84e1cd73929c9935d8381cd7e549199ecf09ff10/torch/cuda/graphs.py#L385-L397) is overly cautious. We don't need it. I am investigating making a new variant of make_graphed_callables that drops this requirement.
3. At least once, a stream fails to join against the capturing stream during stream capture: https://github.com/pytorch/pytorch/blob/1981ed4f60ba54e54373bb44381e301a4b8f594b/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L247 This means this work is not recorded inside the cuda graph, causing a mismatch between graph replay and eager execution.

Unfortunately, several workloads using FSDP2 can become host latency bounded, so cuda graph support is something reasonable to want for it.

I can take care of these issues myself, but I wanted to document them so that people are aware. I also don't personally know the FSDP2 maintainers, so I wanted to make sure they were aware as well.

cc @ezyang @gchanan @kadeng @msaroufim @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @mcarilli @eellison @penguinwu @BoyuanFeng @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @muchulee8 @amjames @chauhang @aakhundov @coconutruben @zou3519

**Comment by weifengpy** [comment]
Hi @galv, fsdp2 maintainer here. thanks for documenting the gaps for fsdp2 + cuda graphs! It your workload cpu bound because of local batch size 1? Is this for finetuning?

**Comment by weifengpy** [comment]
> many workloads on GB200 (aka, Grace CPU with B200 GPU) are host bounded because of the Grace CPU's poor single-threaded performance 

thanks for explaining! this is insightful
